File: scripts/check_path_versions.py
#!/usr/bin/env python3
"""Validate that path dependencies declare versions for publishing."""

from __future__ import annotations

import json
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable, List, Mapping, MutableMapping

ROOT = Path(__file__).resolve().parents[1]


@dataclass
class Problem:
    manifest: Path
    dependency: str
    target: str | None
    path_value: str

    def format(self) -> str:
        location = f"{self.manifest.relative_to(ROOT)}"
        if self.target:
            location = f"{location} ({self.target})"
        return (
            f"{location}: dependency '{self.dependency}' uses path '{self.path_value}' without a version"
        )


def load_workspace_manifests() -> Iterable[Path]:
    metadata = subprocess.run(
        ["cargo", "metadata", "--format-version", "1", "--no-deps"],
        check=True,
        capture_output=True,
        text=True,
        cwd=ROOT,
    )
    data = json.loads(metadata.stdout)
    members = set(data["workspace_members"])
    for package in data["packages"]:
        if package["id"] in members:
            yield Path(package["manifest_path"])


def check_dependency_table(
    table: Mapping[str, object],
    *,
    manifest: Path,
    target: str | None,
    problems: List[Problem],
) -> None:
    for name, raw_spec in table.items():
        if not isinstance(raw_spec, MutableMapping):
            # Specifications that are strings or numbers cannot contain path dependencies.
            continue
        spec = dict(raw_spec)
        if "path" in spec and "version" not in spec:
            problems.append(
                Problem(
                    manifest=manifest,
                    dependency=name,
                    target=target,
                    path_value=str(spec["path"]),
                )
            )


def collect_problems(manifest_path: Path) -> List[Problem]:
    with manifest_path.open("rb") as fh:
        data = tomllib.load(fh)  # type: ignore[name-defined]

    problems: List[Problem] = []
    for section in ("dependencies", "dev-dependencies", "build-dependencies"):
        table = data.get(section)
        if isinstance(table, Mapping):
            check_dependency_table(
                table, manifest=manifest_path, target=None, problems=problems
            )

    target_tables = data.get("target")
    if isinstance(target_tables, Mapping):
        for target_name, target_table in target_tables.items():
            if not isinstance(target_table, Mapping):
                continue
            for section in ("dependencies", "dev-dependencies", "build-dependencies"):
                table = target_table.get(section)
                if isinstance(table, Mapping):
                    check_dependency_table(
                        table,
                        manifest=manifest_path,
                        target=target_name,
                        problems=problems,
                    )

    return problems


try:
    import tomllib  # type: ignore[attr-defined]
except ModuleNotFoundError:  # pragma: no cover - Python <3.11 fallback
    import tomli as tomllib  # type: ignore[no-redef]


def main() -> int:
    problems: List[Problem] = []
    for manifest in load_workspace_manifests():
        problems.extend(collect_problems(manifest))

    if problems:
        for problem in problems:
            print(problem.format(), file=sys.stderr)
        print(
            "error: add an explicit version alongside each path dependency", file=sys.stderr
        )
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check_path_versions.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! anyhow = "1"
//! cargo_metadata = "0.18"
//! toml = { version = "0.8", default-features = false, features = ["parse"] }
//! ```

use std::collections::HashSet;
use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Context, Result};
use cargo_metadata::{MetadataCommand, PackageId};
use toml::value::Table;
use toml::Value;

#[derive(Debug)]
struct Problem {
    manifest: PathBuf,
    dependency: String,
    target: Option<String>,
    path_value: String,
}

impl Problem {
    fn render(&self, root: &Path) -> String {
        let relative = self
            .manifest
            .strip_prefix(root)
            .unwrap_or(self.manifest.as_path());
        let mut location = relative.display().to_string();
        if let Some(target) = &self.target {
            location = format!("{location} ({target})");
        }
        format!(
            "{location}: dependency '{}' uses path '{}' without a version",
            self.dependency, self.path_value
        )
    }
}

fn main() {
    match run() {
        Ok((root, problems)) => {
            if problems.is_empty() {
                return;
            }
            for problem in &problems {
                eprintln!("{}", problem.render(&root));
            }
            eprintln!("error: add an explicit version alongside each path dependency");
            std::process::exit(1);
        }
        Err(error) => {
            eprintln!("error: {error}");
            std::process::exit(1);
        }
    }
}

fn run() -> Result<(PathBuf, Vec<Problem>)> {
    let metadata = MetadataCommand::new()
        .no_deps()
        .exec()
        .context("failed to invoke `cargo metadata`")?;

    let root = metadata.workspace_root.into_std_path_buf();
    let workspace_members: HashSet<PackageId> =
        metadata.workspace_members.iter().cloned().collect();

    let mut problems = Vec::new();
    for package in metadata.packages {
        if !workspace_members.contains(&package.id) {
            continue;
        }
        let manifest_path = package.manifest_path.into_std_path_buf();
        problems.extend(collect_problems(&manifest_path)?);
    }

    Ok((root, problems))
}

fn collect_problems(manifest_path: &Path) -> Result<Vec<Problem>> {
    let contents = fs::read_to_string(manifest_path)
        .with_context(|| format!("failed to read {}", manifest_path.display()))?;
    let parsed: Value = toml::from_str(&contents)
        .with_context(|| format!("failed to parse {}", manifest_path.display()))?;
    let table = parsed
        .as_table()
        .context("manifest root is not a TOML table")?;

    let mut problems = Vec::new();
    for section in ["dependencies", "dev-dependencies", "build-dependencies"] {
        if let Some(deps) = table.get(section).and_then(Value::as_table) {
            check_dependency_table(deps, manifest_path, None, &mut problems);
        }
    }

    if let Some(targets) = table.get("target").and_then(Value::as_table) {
        for (target_name, target_table) in targets {
            if let Some(target_table) = target_table.as_table() {
                for section in ["dependencies", "dev-dependencies", "build-dependencies"] {
                    if let Some(deps) = target_table.get(section).and_then(Value::as_table) {
                        check_dependency_table(
                            deps,
                            manifest_path,
                            Some(target_name.as_str()),
                            &mut problems,
                        );
                    }
                }
            }
        }
    }

    Ok(problems)
}

fn check_dependency_table(
    table: &Table,
    manifest_path: &Path,
    target: Option<&str>,
    problems: &mut Vec<Problem>,
) {
    for (name, spec) in table {
        let Some(spec_table) = spec.as_table() else {
            continue;
        };
        if !spec_table.contains_key("path") || spec_table.contains_key("version") {
            continue;
        }

        let path_value = spec_table
            .get("path")
            .map(value_to_string)
            .unwrap_or_else(|| "<unknown>".to_string());

        problems.push(Problem {
            manifest: manifest_path.to_owned(),
            dependency: name.clone(),
            target: target.map(|t| t.to_string()),
            path_value,
        });
    }
}

fn value_to_string(value: &Value) -> String {
    match value {
        Value::String(s) => s.clone(),
        other => other.to_string(),
   }
}
